src/core/psxinterpreter.cc

```
@@ -997,7 +997,9 @@ void InterpretedCPU::psxSW(uint32_t code) {
 void InterpretedCPU::psxSWL(uint32_t code) {
     uint32_t addr = _oB_;
     uint32_t shift = addr & 3;
-    uint32_t mem = PCSX::g_emulator->m_mem->read32(addr & ~3);
+    // the check avoids msan interpreting a fully aligned store as a read
+    // however it may still interpret misaligned stores as reads
+    uint32_t mem = shift != 3? PCSX::g_emulator->m_mem->read32(addr & ~3): 0;
```

Right, I didn't think about this one with the original version of the code. It might be worth passing an optional mask value to read32 and write32 for the purpose of msan with a default value of 0xffffffff, and then pass the various masks to indicate what's actually being read and written when marking msan's bitmaps.

---

To avoid slowing down read32 since it's much more common than swl/swr, and since those instructions don't actually read, I handled this with a special case in the instruction that doesn't check for the status (then if anything's wrong, the error will come from the following write32)